File: packages/sdk/src/color.ts
/** Full color utility class with hex, rgb, cmyk, hsl support */
export class Color {
  private r: number;
  private g: number;
  private b: number;
  private a: number;

  constructor(r: number, g: number, b: number, a = 1.0) {
    this.r = Math.max(0, Math.min(255, Math.round(r)));
    this.g = Math.max(0, Math.min(255, Math.round(g)));
    this.b = Math.max(0, Math.min(255, Math.round(b)));
    this.a = Math.max(0, Math.min(1, a));
  }

  static hex(hex: string): Color {
    hex = hex.replace('#', '');
    if (hex.length === 3) {
      hex = hex[0]+hex[0]+hex[1]+hex[1]+hex[2]+hex[2];
    }
    const r = parseInt(hex.slice(0, 2), 16);
    const g = parseInt(hex.slice(2, 4), 16);
    const b = parseInt(hex.slice(4, 6), 16);
    const a = hex.length === 8 ? parseInt(hex.slice(6, 8), 16) / 255 : 1;
    return new Color(r, g, b, a);
  }

  static rgb(r: number, g: number, b: number, a = 1.0): Color {
    return new Color(r, g, b, a);
  }

  static hsl(h: number, s: number, l: number, a = 1.0): Color {
    const sn = s / 100, ln = l / 100;
    const c = (1 - Math.abs(2 * ln - 1)) * sn;
    const x = c * (1 - Math.abs((h / 60) % 2 - 1));
    const m = ln - c / 2;
    let [r, g, b] = h < 60 ? [c,x,0] : h < 120 ? [x,c,0] : h < 180 ? [0,c,x] :
      h < 240 ? [0,x,c] : h < 300 ? [x,0,c] : [c,0,x];
    return new Color((r+m)*255, (g+m)*255, (b+m)*255, a);
  }

  static cmyk(c: number, m: number, y: number, k: number): Color {
    const r = 255 * (1-c) * (1-k);
    const g = 255 * (1-m) * (1-k);
    const b = 255 * (1-y) * (1-k);
    return new Color(r, g, b);
  }

  toHex(): string {
    const r = this.r.toString(16).padStart(2, '0');
    const g = this.g.toString(16).padStart(2, '0');
    const b = this.b.toString(16).padStart(2, '0');
    return `#${r}${g}${b}`.toUpperCase();
  }

  toRgb(): { r: number; g: number; b: number; a: number } {
    return { r: this.r, g: this.g, b: this.b, a: this.a };
  }

  toHsl(): { h: number; s: number; l: number } {
    const r = this.r / 255, g = this.g / 255, b = this.b / 255;
    const max = Math.max(r, g, b), min = Math.min(r, g, b);
    const l = (max + min) / 2;
    if (max === min) return { h: 0, s: 0, l: l * 100 };
    const d = max - min;
    const s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
    let h = max === r ? (g-b)/d+(g<b?6:0) : max === g ? (b-r)/d+2 : (r-g)/d+4;
    return { h: h * 60, s: s * 100, l: l * 100 };
  }

  toCmyk(): { c: number; m: number; y: number; k: number } {
    const r = this.r / 255, g = this.g / 255, b = this.b / 255;
    const k = 1 - Math.max(r, g, b);
    if (k >= 1) return { c: 0, m: 0, y: 0, k: 1 };
    return {
      c: (1 - r - k) / (1 - k),
      m: (1 - g - k) / (1 - k),
      y: (1 - b - k) / (1 - k),
      k,
    };
  }

  withAlpha(a: number): Color {
    return new Color(this.r, this.g, this.b, a);
  }

  // Preset colors
  static readonly black = Color.hex('#000000');
  static readonly white = Color.hex('#FFFFFF');
  static readonly red = Color.hex('#FF0000');
  static readonly green = Color.hex('#00FF00');
  static readonly blue = Color.hex('#0000FF');
  static readonly yellow = Color.hex('#FFFF00');
  static readonly cyan = Color.hex('#00FFFF');
  static readonly magenta = Color.hex('#FF00FF');
  static readonly transparent = new Color(0, 0, 0, 0);
}

File: packages/core/src/pdf/color.rs
/// Color representation with multiple color space support
#[derive(Debug, Clone, PartialEq)]
pub enum Color {
    Rgb(f64, f64, f64),           // 0.0-1.0
    Cmyk(f64, f64, f64, f64),     // 0.0-1.0
    Grayscale(f64),                // 0.0-1.0
}

impl Color {
    /// Parse hex color string: #RGB, #RRGGBB, #RRGGBBAA
    pub fn from_hex(hex: &str) -> Option<Self> {
        let hex = hex.trim_start_matches('#');
        let bytes = match hex.len() {
            3 => {
                let r = u8::from_str_radix(&hex[0..1], 16).ok()? * 17;
                let g = u8::from_str_radix(&hex[1..2], 16).ok()? * 17;
                let b = u8::from_str_radix(&hex[2..3], 16).ok()? * 17;
                (r, g, b)
            }
            6 | 8 => {
                let r = u8::from_str_radix(&hex[0..2], 16).ok()?;
                let g = u8::from_str_radix(&hex[2..4], 16).ok()?;
                let b = u8::from_str_radix(&hex[4..6], 16).ok()?;
                (r, g, b)
            }
            _ => return None,
        };
        Some(Color::Rgb(
            bytes.0 as f64 / 255.0,
            bytes.1 as f64 / 255.0,
            bytes.2 as f64 / 255.0,
        ))
    }

    pub fn from_rgb(r: u8, g: u8, b: u8) -> Self {
        Color::Rgb(r as f64 / 255.0, g as f64 / 255.0, b as f64 / 255.0)
    }

    pub fn from_cmyk(c: f64, m: f64, y: f64, k: f64) -> Self {
        Color::Cmyk(c, m, y, k)
    }

    pub fn from_hsl(h: f64, s: f64, l: f64) -> Self {
        let s = s / 100.0;
        let l = l / 100.0;
        let c = (1.0 - (2.0 * l - 1.0).abs()) * s;
        let x = c * (1.0 - ((h / 60.0) % 2.0 - 1.0).abs());
        let m = l - c / 2.0;
        let (r, g, b) = match h as u32 {
            0..=59 => (c, x, 0.0),
            60..=119 => (x, c, 0.0),
            120..=179 => (0.0, c, x),
            180..=239 => (0.0, x, c),
            240..=299 => (x, 0.0, c),
            _ => (c, 0.0, x),
        };
        Color::Rgb(r + m, g + m, b + m)
    }

    pub fn to_rgb(&self) -> (f64, f64, f64) {
        match self {
            Color::Rgb(r, g, b) => (*r, *g, *b),
            Color::Cmyk(c, m, y, k) => {
                let r = (1.0 - c) * (1.0 - k);
                let g = (1.0 - m) * (1.0 - k);
                let b = (1.0 - y) * (1.0 - k);
                (r, g, b)
            }
            Color::Grayscale(g) => (*g, *g, *g),
        }
    }

    pub fn to_hex(&self) -> String {
        let (r, g, b) = self.to_rgb();
        format!("#{:02X}{:02X}{:02X}",
            (r * 255.0) as u8,
            (g * 255.0) as u8,
            (b * 255.0) as u8,
        )
    }

    pub fn to_cmyk(&self) -> (f64, f64, f64, f64) {
        let (r, g, b) = self.to_rgb();
        let k = 1.0 - r.max(g).max(b);
        if k >= 1.0 {
            return (0.0, 0.0, 0.0, 1.0);
        }
        let c = (1.0 - r - k) / (1.0 - k);
        let m = (1.0 - g - k) / (1.0 - k);
        let y = (1.0 - b - k) / (1.0 - k);
        (c, m, y, k)
    }

    /// Write PDF color-setting operators for stroke
    pub fn write_stroke_op(&self, buf: &mut Vec<u8>) {
        match self {
            Color::Rgb(r, g, b) => {
                let s = format!("{} {} {} RG\n", fmt(*r), fmt(*g), fmt(*b));
                buf.extend_from_slice(s.as_bytes());
            }
            Color::Cmyk(c, m, y, k) => {
                let s = format!("{} {} {} {} K\n", fmt(*c), fmt(*m), fmt(*y), fmt(*k));
                buf.extend_from_slice(s.as_bytes());
            }
            Color::Grayscale(g) => {
                let s = format!("{} G\n", fmt(*g));
                buf.extend_from_slice(s.as_bytes());
            }
        }
    }

    /// Write PDF color-setting operators for fill
    pub fn write_fill_op(&self, buf: &mut Vec<u8>) {
        match self {
            Color::Rgb(r, g, b) => {
                let s = format!("{} {} {} rg\n", fmt(*r), fmt(*g), fmt(*b));
                buf.extend_from_slice(s.as_bytes());
            }
            Color::Cmyk(c, m, y, k) => {
                let s = format!("{} {} {} {} k\n", fmt(*c), fmt(*m), fmt(*y), fmt(*k));
                buf.extend_from_slice(s.as_bytes());
            }
            Color::Grayscale(g) => {
                let s = format!("{} g\n", fmt(*g));
                buf.extend_from_slice(s.as_bytes());
            }
        }
    }

    // Common colors
    pub fn black() -> Self { Color::Grayscale(0.0) }
    pub fn white() -> Self { Color::Grayscale(1.0) }
    pub fn red() -> Self { Color::Rgb(1.0, 0.0, 0.0) }
    pub fn green() -> Self { Color::Rgb(0.0, 1.0, 0.0) }
    pub fn blue() -> Self { Color::Rgb(0.0, 0.0, 1.0) }
    pub fn transparent() -> Self { Color::Rgb(0.0, 0.0, 0.0) } // needs separate opacity
}

fn fmt(v: f64) -> String {
    if v == v.floor() && v.abs() < 1e10 {
        format!("{:.1}", v)
    } else {
        format!("{:.4}", v).trim_end_matches('0').trim_end_matches('.').to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hex_parse() {
        let c = Color::from_hex("#FF0000").unwrap();
        assert_eq!(c, Color::Rgb(1.0, 0.0, 0.0));
    }

    #[test]
    fn test_hex_short() {
        let c = Color::from_hex("#F00").unwrap();
        assert_eq!(c, Color::Rgb(1.0, 0.0, 0.0));
    }

    #[test]
    fn test_to_hex() {
        assert_eq!(Color::red().to_hex(), "#FF0000");
    }
}
